src/core/src/signature.rs

```diff
@@ -20,138 +20,279 @@ use crate::sketch::Sketch;
+
+                if self.hash_function.protein() {
+                    let aa_kmer = &self.sequence[self.kmer_index..self.kmer_index + self.k_size];
+                    let hash = crate::_hash_murmur(aa_kmer, self.seed);
+                    self.kmer_index += 1;
+                    Some(Ok(hash))
+                } else {
+                    if !self._prot_configured {
+                        self._aa_seq = match self.hash_function {
+                            HashFunctions::murmur64_dayhoff => {
+                                self.sequence.iter().cloned().map(aa_to_dayhoff).collect()
+                            }
+                            HashFunctions::murmur64_hp => {
+                                self.sequence.iter().cloned().map(aa_to_hp).collect()
+                            }
+                            invalid => {
+                                return Some(Err(Error::InvalidHashFunction {
+                                    function: format!("{}", invalid),
+                                }));
+                            }
+                        };
+                    }
 
-                let krc = &rc[len - ksize - i..len - i];
-                let hash = crate::_hash_murmur(std::cmp::min(kmer, krc), self.seed());
-                self.add_hash(hash);
+                    let aa_kmer = &self._aa_seq[self.kmer_index..self.kmer_index + self.k_size];
+                    let hash = crate::_hash_murmur(aa_kmer, self.seed);
+                    self.kmer_index += 1;
+                    Some(Ok(hash))
+                }
             }
         } else {
-            // protein
-            let aa_ksize = self.ksize() / 3;
-
-            for i in 0..3 {
-                let substr: Vec<u8> = sequence
-                    .iter()
-                    .cloned()
-                    .skip(i)
-                    .take(sequence.len() - i)
-                    .collect();
-                let aa = to_aa(&substr, hash_function.dayhoff(), hash_function.hp()).unwrap();
-
-                aa.windows(aa_ksize as usize).for_each(|n| {
-                    let hash = crate::_hash_murmur(n, self.seed());
-                    self.add_hash(hash);
-                });
-
-                let rc_substr: Vec<u8> = rc.iter().cloned().skip(i).take(rc.len() - i).collect();
-                let aa_rc = to_aa(&rc_substr, hash_function.dayhoff(), hash_function.hp()).unwrap();
-
-                aa_rc.windows(aa_ksize as usize).for_each(|n| {
-                    let hash = crate::_hash_murmur(n, self.seed());
-                    self.add_hash(hash);
-                });
-            }
+            // End the iterator
+            None
         }
-
-        Ok(())
     }
+}
 
-    fn add_protein(&mut self, seq: &[u8]) -> Result<(), Error> {
-        let ksize = (self.ksize() / 3) as usize;
-        let len = seq.len();
-        let hash_function = self.hash_function();
+pub trait SigsTrait {
```

Comment:
minor nitpick: can we keep `SigsTrait` on the top? There are no hard rules, but trait/enum/struct declarations tend to come before implementations.

Reply:
Resolved in [`9bed5d9` (#1653)](https://github.com/sourmash-bio/sourmash/pull/1653/commits/9bed5d9cf448639647c159562cb585b12b505f6b)